modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java

```diff
@@ -4826,68 +4778,60 @@ public void awaitPendingTasksFinished() throws IgniteCheckedException {
+        for (T2<PageMemoryEx, GridMultiCollectionWrapper<FullPageId>> regPages : cpPages.get1()) {
+            FullPageId[] pages = new FullPageId[regPages.get2().size()];
 
-        // Some pages may have been replaced, need to fill end of array with fake ones to prevent NPE during sort.
-        for (int i = realPagesArrSize; i < pagesArr.length; i++)
-            pagesArr[i] = fakeMaxFullPageId;
+            int pagePos = 0;
 
-        if (persistenceCfg.getCheckpointWriteOrder() == CheckpointWriteOrder.SEQUENTIAL) {
-            Comparator<FullPageId> cmp = new Comparator<FullPageId>() {
-                @Override public int compare(FullPageId o1, FullPageId o2) {
-                    int cmp = Long.compare(o1.groupId(), o2.groupId());
-                    if (cmp != 0)
-                        return cmp;
+            for (int i = 0; i < regPages.get2().collectionsSize(); i++) {
+                for (FullPageId page : regPages.get2().innerCollection(i)) {
+                    if (realPagesArrSize++ == totalPagesCnt)
+                        throw new AssertionError("Incorrect estimated dirty pages number: " + cpPages.get2());
 
-                    return Long.compare(o1.effectivePageId(), o2.effectivePageId());
+                    pages[pagePos++] = page;
                 }
-            };
+            }
 
-            if (pagesArr.length >= parallelSortThreshold)
-                parallelSortInIsolatedPool(pagesArr, cmp);
+            // Some pages may have been already replaced.
+            if (pagePos != pages.length)
+                cpPagesPerRegion.add(new T2<>(regPages.get1(), Arrays.copyOf(pages, pagePos)));
             else
-                Arrays.sort(pagesArr, cmp);
+                cpPagesPerRegion.add(new T2<>(regPages.get1(), pages));
         }
 
-        int pagesSubLists = threads == 1 ? 1 : threads * 4;
-        // Splitting pages to (threads * 4) subtasks. If any thread will be faster, it will help slower threads.
-
-        Collection[] pagesSubListArr = new Collection[pagesSubLists];
-
-        for (int i = 0; i < pagesSubLists; i++) {
-            int from = (int)((long)realPagesArrSize * i / pagesSubLists);
+        if (persistenceCfg.getCheckpointWriteOrder() == CheckpointWriteOrder.SEQUENTIAL) {
+            Comparator<FullPageId> cmp = (o1, o2) -> {
+                int cmp1 = Long.compare(o1.groupId(), o2.groupId());
+                if (cmp1 != 0)
+                    return cmp1;
 
-            int to = (int)((long)realPagesArrSize * (i + 1) / pagesSubLists);
+                return Long.compare(o1.effectivePageId(), o2.effectivePageId());
+            };
 
-            pagesSubListArr[i] = new GridReadOnlyArrayView(pagesArr, from, to);
+            for (T2<PageMemoryEx, FullPageId[]> pagesPerReg : cpPagesPerRegion) {
+                if (pagesPerReg.getValue().length >= parallelSortThreshold)
+                    parallelSortInIsolatedPool(pagesPerReg.get2(), cmp);
```

Comment:
This method will create new pool for every region the exceeds its threshold, but we can reuse it.

You see, this logic was very simple before the refactoring. Single array was already sorted by groupId, so resorting was not super heavy. Do you have some benchmarks showing that this loop actually helps?

Reply:
refactor it, plz check.